File: WebAssemblyBrowserApp/WebAssembly_WebGPU/wwwroot/jsengine/audio/audio.js
// =====================================================================
// 音效层：WebAudio 简单合成（方波/正弦 beep）。
// 音频上下文须由用户手势解锁（ensure()），在输入事件中调用。
// =====================================================================

let _audioCtx = null

export const audio = {
    init() { _audioCtx = null },

    ensure() {
        if (!_audioCtx) {
            const AC = window.AudioContext || window.webkitAudioContext
            if (AC) _audioCtx = new AC()
        }
        if (_audioCtx && _audioCtx.state === 'suspended') _audioCtx.resume()
    },

    beep(freq, dur, waveType, vol) {
        if (!_audioCtx) return
        try {
            const osc = _audioCtx.createOscillator()
            const gain = _audioCtx.createGain()
            osc.type = waveType || 'square'
            osc.frequency.value = freq
            const t = _audioCtx.currentTime
            const v = vol || 0.08
            gain.gain.setValueAtTime(v, t)
            gain.gain.exponentialRampToValueAtTime(0.0001, t + (dur || 0.1))
            osc.connect(gain)
            gain.connect(_audioCtx.destination)
            osc.start(t)
            osc.stop(t + (dur || 0.1) + 0.02)
        } catch { /* 忽略音频异常 */ }
    },
}

File: WebAssemblyBrowserApp/WebEngineCommon/Engine/Audio.cs
using System.Runtime.InteropServices.JavaScript;

namespace WebAssemblyBrowserApp.Engine;

/// <summary>基于 WebAudio 的简单音效（JS 薄层封装）。</summary>
public static partial class Audio
{
    [JSImport("audio.init", "main.js")]
    public static partial void Init();

    /// <summary>确保 AudioContext 已创建并在用户手势后恢复。</summary>
    [JSImport("audio.ensure", "main.js")]
    public static partial void Ensure();

    /// <summary>播放一个短音。waveType: sine / square / triangle / sawtooth。</summary>
    [JSImport("audio.beep", "main.js")]
    public static partial void Beep(double frequency, double duration, string waveType, double volume);
}
